### ISSUE TITLE
Allow escaping of special characters.

### ISSUE THREAD

**Issue by Kevinpgalligan** [Kevinpgalligan/KotlinRegex#2]
You should be allowed to escape special characters, e.g. "\(" should be treated as "(" rather than the opening symbol for groups.

I think this change will involve only the Tokenizer. Rather than returning a special token, return a RawCharacter token. It'll require a bit of refactoring since the Tokenizer currently maps 1 character -> 1 token, we will need something slightly more intelligent.

**Comment by j-jzk** [comment]
I think that in combination with #3 it would be better to put this into the parser, since it would need to be context-aware. The tokenizer would need to know if we wanted to escape a character (`\(`) or add a substitution (`\w`), which is in the scope of the parser IMO.

**Comment by j-jzk** [comment]
Actually, after implementing #15, I've come to the conclusion that implementing functionality like I described above would quickly turn the code into a spaghettified mess. I think what we need is for the ~~parser~~ **tokenizer** to have some sorts of "modes", possibly triggered by different characters. 
For example, when tokenizing a caret inside a character range (`[^abc]`), we would want the caret to be encoded as a special character, but otherwise not. Character escaping could be implemented similarly.
I will think this through; ideas welcome.

**Comment by Kevinpgalligan** [comment]
Agreed that it makes more sense to put this in the parser! Putting it in the tokenizer seems like a hack.

I think groups, and character ranges, and escape sequences, can probably be incorporated into the grammar & the parser. The rule for a character range might look something like: `CHAR_RANGE -> '[' CHAR_RANGE_INVERT? CHAR_RANGE_ITEMS+ ']'`, which could be implemented in the parser using recursive descent. Escape sequences might be incorporated into the rule for regular characters: `CHAR -> 'a' | ... | 'z' | ESCAPE_MARKER ESCAPED_CHAR` and `ESCAPED_CHAR -> '(' | ')' | ...`.

How you would take the resulting parse object and turn it into a pattern matching thingy is another question. Character ranges can be represented as finite automata. Groups can be stripped out and stored as meta information, perhaps? And escape characters can be treated like regular characters.

It has been ages since I even looked at this code, though, so I might be disconnected from reality, haha.

**Comment by j-jzk** [comment]
What I was talking about was that when we tokenize the regex, we assign special meaning to some characters (like parentheses), but this is problematic for some (like a caret) because they have double meaning.

If we used a separate singleton object for them, we would need to add special cases everywhere where they are to be treated as a regular character. Tokenizing them like regular characters and then adding special cases into the parts where they have special meaning *feels* like a hack, but might actually be a nice solution.

For the actual inner working of the character classes/ranges, we can just simply make them a CharMatcher with a new implementation of Symbol. I don't intend to implement group capture as of now, so that's not a thing to worry about.
